pkg/pack/resources.go

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright 2020 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pack
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	kyaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	"github.com/crossplane/crossplane/apis/workload/v1alpha1"
+)
+
+// ReadResources reads Kubernetes objects and packages them into a slice of
+// KubernetesApplicationResourceTemplates.
+func ReadResources(prefix string, content io.Reader, resLabels map[string]string) ([]v1alpha1.KubernetesApplicationResourceTemplate, error) {
+	var result []v1alpha1.KubernetesApplicationResourceTemplate
+	d := kyaml.NewYAMLOrJSONDecoder(content, 4096)
+	for {
+		obj := &unstructured.Unstructured{}
+		if err := d.Decode(obj); err != nil {
+			if err == io.EOF {
+				// We reached the end of the job output
+				break
+			}
+			return result, err
+		}
+		if obj.GetName() == "" {
+			continue
+		}
+		kart := v1alpha1.KubernetesApplicationResourceTemplate{
+			ObjectMeta: v1.ObjectMeta{
+				// Ensure no collisions among templates in KubernetesApplications
+				Name:   fmt.Sprintf("%s-local-%s-%s", prefix, strings.ReplaceAll(obj.GetName(), ":", "-"), strings.ToLower(obj.GetKind())),
```

Comment:
This is one of the things I hate about writing `KubernetesApplication`; a separate name for KAR objects that you have to think about. I just put `-local-` for the POC purposes, but I believe it looks better if we skip that. Something like:
```suggestion
				Name:   fmt.Sprintf("%s-%s", prefix, strings.ReplaceAll(obj.GetName(), ":", "-"),
```

We know that it's `local` since it's KAR, namespace is the same and kind is already in the spec as well as printer columns. So, `KAPNAME-OBJNAME` seemed like a better UX to me. It's also better to keep it short to not hit name limit. Though all this is nitpick.

Reply:
@muvaf I think we can remove `local` but we need to include the object `Kind` or else we can have collisions with objects of different `Kind` but same `Name` because the KAR they are packaged in will have the same name.